File: mtgSearch/src/main/java/com/dbottillo/mtgsearchfree/database/SetDataSource.java
package com.dbottillo.mtgsearchfree.database;

import android.content.ContentValues;
import android.database.Cursor;
import android.database.sqlite.SQLiteDatabase;

import com.dbottillo.mtgsearchfree.model.MTGSet;

import org.json.JSONException;
import org.json.JSONObject;

import java.util.ArrayList;

public final class SetDataSource {

    public static final String TABLE = "MTGSet";

    public enum COLUMNS {
        NAME("name", "TEXT"),
        CODE("code", "TEXT");

        private String name;
        private String type;

        COLUMNS(String name, String type) {
            this.name = name;
            this.type = type;
        }

        public String getName() {
            return name;
        }
    }

    private SetDataSource() {
    }

    public static String generateCreateTable() {
        StringBuilder builder = new StringBuilder("CREATE TABLE IF NOT EXISTS ");
        builder.append(TABLE).append(" (_id INTEGER PRIMARY KEY, ");
        for (COLUMNS column : COLUMNS.values()) {
            builder.append(column.name).append(' ').append(column.type);
            if (column != COLUMNS.CODE) {
                builder.append(',');
            }
        }
        return builder.append(')').toString();
    }

    public static long saveSet(SQLiteDatabase db, MTGSet set) {
        ContentValues values = new ContentValues();
        if (set.getId() > -1) {
            values.put("_id", set.getId());
        }
        values.put(COLUMNS.NAME.getName(), set.getName());
        values.put(COLUMNS.CODE.getName(), set.getCode());
        return db.insertWithOnConflict(TABLE, null, values, SQLiteDatabase.CONFLICT_REPLACE);
    }

    public static ArrayList<MTGSet> getSets(SQLiteDatabase db) {
        String query = "SELECT * FROM " + TABLE;
        Cursor cursor = db.rawQuery(query, null);
        ArrayList<MTGSet> sets = new ArrayList<>();
        if (cursor.moveToFirst()) {
            while (!cursor.isAfterLast()) {
                sets.add(fromCursor(cursor));
                cursor.moveToNext();
            }
        }
        cursor.close();
        return sets;
    }

    public static void removeSet(SQLiteDatabase database, long id) {
        String[] args = new String[]{id + ""};
        database.rawQuery("DELETE FROM " + TABLE + " where _id=? ", args).moveToFirst();
    }

    public static MTGSet fromCursor(Cursor cursor) {
        MTGSet set = new MTGSet(cursor.getInt(cursor.getColumnIndex("_id")));
        set.setName(cursor.getString(cursor.getColumnIndex(COLUMNS.NAME.getName())));
        set.setCode(cursor.getString(cursor.getColumnIndex(COLUMNS.CODE.getName())));
        return set;
    }

    public static ContentValues fromJSON(JSONObject object) throws JSONException {
        ContentValues values = new ContentValues();
        values.put(COLUMNS.CODE.getName(), object.getString("code"));
        values.put(COLUMNS.NAME.getName(), object.getString("name"));
        return values;
    }
}

File: data_sets/src/main/kotlin/com/dbottillo/mtgsearchfree/storage/SetDataSource.kt
package com.dbottillo.mtgsearchfree.storage

import android.content.ContentValues
import android.database.Cursor
import android.database.sqlite.SQLiteDatabase
import com.dbottillo.mtgsearchfree.model.MTGSet
import com.dbottillo.mtgsearchfree.model.SetType
import com.dbottillo.mtgsearchfree.model.SetType.COMMANDER
import com.dbottillo.mtgsearchfree.model.SetType.EXPANSION
import com.dbottillo.mtgsearchfree.model.SetType.FUNNY
import com.dbottillo.mtgsearchfree.model.SetType.PREVIEW
import com.dbottillo.mtgsearchfree.model.SetType.PROMO
import com.dbottillo.mtgsearchfree.util.LOG
import com.dbottillo.mtgsearchfree.util.add
import com.dbottillo.mtgsearchfree.util.fromJson
import com.dbottillo.mtgsearchfree.util.getIntFromColumn
import com.dbottillo.mtgsearchfree.util.getStringFromColumn
import org.json.JSONException
import org.json.JSONObject

class SetDataSource(private val database: SQLiteDatabase) {

    val sets: List<MTGSet>
        get() {
            val query = "SELECT * FROM $TABLE"
            val cursor = database.rawQuery(query, null)
            val sets = ArrayList<MTGSet>()
            if (cursor.moveToFirst()) {
                while (!cursor.isAfterLast) {
                    sets.add(fromCursor(cursor))
                    cursor.moveToNext()
                }
            }
            LOG.query(query)
            cursor.close()
            return sets
        }

    fun saveSet(set: MTGSet): Long {
        val values = ContentValues().apply {
            if (set.id > -1) {
                put("_id", set.id)
            }
            put("name", set.name)
            put("code", set.code)
            put("type", wrap(set.type))
        }
        return database.insertWithOnConflict(TABLE, null, values, SQLiteDatabase.CONFLICT_REPLACE)
    }

    fun removeSet(id: Long) {
        val args = arrayOf(id.toString() + "")
        val query = "DELETE FROM $TABLE where _id=? "
        val cursor = database.rawQuery(query, args)
        cursor.moveToFirst()
        cursor.close()
        LOG.query(query, *args)
    }

    fun fromCursor(cursor: Cursor): MTGSet {
        val type = cursor.getString(cursor.getColumnIndex("type")) ?: null
        return MTGSet(id = cursor.getIntFromColumn("_id"),
                name = cursor.getStringFromColumn("name"),
                code = cursor.getStringFromColumn("code"),
                type = unwrap(type))
    }

    fun unwrap(input: String?): SetType {
        return when (input) {
            "preview" -> PREVIEW
            "funny" -> FUNNY
            "commander" -> COMMANDER
            "promo" -> PROMO
            else -> EXPANSION
        }
    }

    private fun wrap(input: SetType): String {
        return when (input) {
            PREVIEW -> "preview"
            FUNNY -> "funny"
            else -> "normal"
        }
    }

    @Throws(JSONException::class)
    fun fromJSON(jsonObject: JSONObject): ContentValues {
        return ContentValues().apply {
            fromJson("name", jsonObject)
            fromJson("code", jsonObject)
            fromJson("type", jsonObject)
        }
    }

    companion object {

        const val TABLE = "MTGSet"

        fun generateCreateTable(): String {
            val builder = StringBuilder("CREATE TABLE IF NOT EXISTS ")
            builder.append(TABLE).append(" (_id INTEGER PRIMARY KEY, ")
            builder.add(name = "name", type = "TEXT")
            builder.add(name = "code", type = "TEXT")
            builder.add(name = "type", type = "TEXT", last = true)
            return builder.append(')').toString()
        }
    }
}